contrib/hadoop-store-builder/src/java/voldemort/store/readonly/disk/HadoopStoreWriter.java

```diff
@@ -213,13 +226,33 @@ public void write(BytesWritable key, Iterator<BytesWritable> iterator, Reporter
             int offsetTillNow = 0;
 
             // Read node Id
-            if(this.nodeId == -1)
-                this.nodeId = ByteUtils.readInt(valueBytes, offsetTillNow);
+            int currentNodeId = ByteUtils.readInt(valueBytes, offsetTillNow);
+            if (this.nodeId == -1) {
+                this.nodeId = currentNodeId;
+            } else if (this.nodeId != currentNodeId) {
+                /**
+                 * Defensive code to catch regressions in
+                 * {@link AbstractStoreBuilderConfigurable#getPartition(byte[], byte[], int)}
+                 */
+                throw new IllegalArgumentException("Should get various nodeId shuffled to us! "
+                                                   + "First nodeId seen: " + this.nodeId
+                                                   + ", currentNodeId: " + currentNodeId);
+            }
             offsetTillNow += ByteUtils.SIZE_OF_INT;
 
             // Read partition id
-            if(this.partitionId == -1)
-                this.partitionId = ByteUtils.readInt(valueBytes, offsetTillNow);
+            int currentPartitionId = ByteUtils.readInt(valueBytes, offsetTillNow);
+            if (this.partitionId == -1) {
+                this.partitionId = currentPartitionId;
+            } else if (this.partitionId != currentPartitionId) {
```

Comment:
we should also validate the replica type. The error message sounds little confusing.